pip/req/req_uninstall.py

```diff
@@ -131,60 +131,78 @@ def compact(self, paths):
         necessary to contain all paths in the set. If /a/path/ and
         /a/path/to/a/file.txt are both in the set, leave only the
         shorter path."""
+
+        sep = os.path.sep
         short_paths = set()
         for path in sorted(paths, key=len):
-            if not any([
-                    (path.startswith(shortpath) and
-                     path[len(shortpath.rstrip(os.path.sep))] == os.path.sep)
-                    for shortpath in short_paths]):
+            should_add = any(
+                path.startswith(shortpath.rstrip("*")) and
+                path[len(shortpath.rstrip("*").rstrip(sep))] == sep
+                for shortpath in short_paths
+            )
+            if not should_add:
```

Comment:
That sounds reasonable. If there are still cases where it's not compact enough, we can review those to work out what additional improvements are practical and safe. (We may need to special-case ```__pycache__```, although https://github.com/pypa/pip/blob/master/pip/req/req_uninstall.py#L115 might cover it)